File: mobile/modules/jsc/ios/JscModule.swift
import ExpoModulesCore
import Foundation

public class JscModule: Module {
  static weak var shared: JscModule?

  public func definition() -> ModuleDefinition {
    Name("Jsc")

    Events("foreground_message", "jsc_console", "jsc_error")

    OnCreate {
      JscModule.shared = self
      JscRuntime.shared.bind(module: self)
    }

    OnDestroy {
      JscRuntime.shared.destroyAll()
      JscModule.shared = nil
    }

    AsyncFunction("createContext") { (id: String) in
      try JscRuntime.shared.create(id: id)
    }

    AsyncFunction("destroyContext") { (id: String) in
      JscRuntime.shared.destroy(id: id)
    }

    AsyncFunction("evaluate") { (id: String, script: String) -> Any? in
      try JscRuntime.shared.evaluate(id: id, script: script)
    }

    Function("listContexts") { () -> [String] in
      JscRuntime.shared.listIds()
    }
  }

  /// Internal: emit a typed event with a payload dict. Always called from
  /// JscRuntime/JscBridge — never directly from JS.
  func emit(_ name: String, _ payload: [String: Any]) {
    sendEvent(name, payload)
  }
}

File: mobile/modules/jsc/android/src/main/java/com/mentra/jsc/JscModule.kt
package com.mentra.jsc

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android stub for @mentra/jsc.
 *
 * The full JSContext-based runtime lives only on iOS (see ios/Source/JscRuntime.swift).
 * On Android we expose the same TS surface but every call rejects, so imports
 * type-check and consumers can no-op feature-detect via `try { await Jsc.createContext(...) }`.
 */
class JscModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Jsc")

    Events("foreground_message", "jsc_console", "jsc_error")

    AsyncFunction("createContext") { _: String ->
      throw NotSupportedException()
    }

    AsyncFunction("destroyContext") { _: String ->
      // No-op: nothing to destroy on Android.
    }

    AsyncFunction("evaluate") { _: String, _: String ->
      throw NotSupportedException()
    }

    Function("listContexts") {
      emptyList<String>()
    }
  }
}

private class NotSupportedException :
  CodedException("Jsc: this module is iOS only; no JSContext runtime on Android.")
